Distribution/Package/ModuleForest.hs

```diff
@@ -0,0 +1,41 @@
+-- Adopted from https://github.com/haskell/hackage-server/blob/master/Distribution/Server/Packages/ModuleForest.hs
+
+module Distribution.Package.ModuleForest
+    ( moduleName
+    , moduleForest
+    , ModuleTree(..)
+    , ModuleForest
+    , NameComponent
+    ) where
+
+import           Distribution.ModuleName (ModuleName)
+import qualified Distribution.ModuleName as ModuleName
+import           Import
+
+type NameComponent = Text
+
+type ModuleForest = [ModuleTree]
+data ModuleTree = Node { component  :: NameComponent
+                       , isModule   :: Bool
+                       , subModules :: ModuleForest
+                       }
+                deriving (Show, Eq)
+
+moduleName :: Text -> ModuleName
+moduleName = ModuleName.fromString . unpack
```

Comment:
Something to be cautious about. `ModuleName.fromString` is defined only for valid module names. Would it be safe to assume that all module names in the database are valid?